File: reference/lexer.py
from tokens import Token, TokenKind, Span, KEYWORDS
from errors import LexError
from typing import List


class Lexer:
    def __init__(self, source: str, filename: str = "<input>"):
        self.source = source
        self.filename = filename
        self.pos = 0
        self.line = 1
        self.col = 1
        self.tokens: List[Token] = []

    def peek(self, offset: int = 0) -> str:
        p = self.pos + offset
        if p < len(self.source):
            return self.source[p]
        return "\0"

    def advance(self) -> str:
        ch = self.source[self.pos]
        self.pos += 1
        if ch == "\n":
            self.line += 1
            self.col = 1
        else:
            self.col += 1
        return ch

    def skip_whitespace(self) -> None:
        while self.pos < len(self.source) and self.source[self.pos] in " \t\r\n":
            self.advance()

    def skip_comment(self) -> bool:
        if self.peek() == "/" and self.peek(1) == "/":
            while self.pos < len(self.source) and self.peek() != "\n":
                self.advance()
            return True
        if self.peek() == "/" and self.peek(1) == "*":
            self.advance()
            self.advance()
            depth = 1
            while self.pos < len(self.source) and depth > 0:
                if self.peek() == "/" and self.peek(1) == "*":
                    self.advance()
                    self.advance()
                    depth += 1
                elif self.peek() == "*" and self.peek(1) == "/":
                    self.advance()
                    self.advance()
                    depth -= 1
                else:
                    self.advance()
            return True
        return False

    def read_string(self) -> Token:
        start_line, start_col = self.line, self.col
        self.advance()  # skip opening "
        value = ""
        while self.pos < len(self.source) and self.peek() != '"':
            if self.peek() == "\\":
                self.advance()
                escape = self.advance()
                escape_map = {"n": "\n", "t": "\t", "r": "\r", "0": "\\", "\\": "\\"}
                value += escape_map.get(escape, escape)
            else:
                value += self.advance()
        if self.pos >= len(self.source):
            raise LexError("unterminated string literal", Span(start_line, start_col), self.filename)
        self.advance()  # skip closing "
        return Token(TokenKind.STRING, value, Span(start_line, start_col, len(value)))

    def read_char(self) -> Token:
        start_line, start_col = self.line, self.col
        self.advance()  # skip opening '
        if self.peek() == "\\":
            self.advance()
            escape = self.advance()
            escape_map = {"n": "\n", "t": "\t", "r": "\r", "0": "\\", "\\": "\\"}
            ch = escape_map.get(escape, escape)
        else:
            ch = self.advance()
        if self.peek() != "'":
            raise LexError("unterminated character literal", Span(start_line, start_col), self.filename)
        self.advance()  # skip closing '
        return Token(TokenKind.CHAR, ch, Span(start_line, start_col, len(ch)))

    def read_number(self) -> Token:
        start_line, start_col = self.line, self.col
        value = ""
        is_float = False

        if self.peek() == "0" and self.peek(1) in "xX":
            value += self.advance()
            value += self.advance()
            while self.pos < len(self.source) and self.peek() in "0123456789abcdefABCDEF_":
                value += self.advance()
            return Token(TokenKind.INT, value, Span(start_line, start_col, len(value)))

        if self.peek() == "0" and self.peek(1) in "bB":
            value += self.advance()
            value += self.advance()
            while self.pos < len(self.source) and self.peek() in "01_":
                value += self.advance()
            return Token(TokenKind.INT, value, Span(start_line, start_col, len(value)))

        if self.peek() == "0" and self.peek(1) in "oO":
            value += self.advance()
            value += self.advance()
            while self.pos < len(self.source) and self.peek() in "01234567_":
                value += self.advance()
            return Token(TokenKind.INT, value, Span(start_line, start_col, len(value)))

        while self.pos < len(self.source) and self.peek().isdigit():
            value += self.advance()

        if self.peek() == "_" or (self.peek().isdigit()):
            while self.pos < len(self.source) and (self.peek().isdigit() or self.peek() == "_"):
                value += self.advance()

        if self.peek() == ".":
            is_float = True
            value += self.advance()
            while self.pos < len(self.source) and (self.peek().isdigit() or self.peek() == "_"):
                value += self.advance()

        if self.peek() in "eE":
            is_float = True
            value += self.advance()
            if self.peek() in "+-":
                value += self.advance()
            while self.pos < len(self.source) and (self.peek().isdigit() or self.peek() == "_"):
                value += self.advance()

        kind = TokenKind.FLOAT if is_float else TokenKind.INT
        return Token(kind, value, Span(start_line, start_col, len(value)))

    def read_ident(self) -> Token:
        start_line, start_col = self.line, self.col
        value = ""
        while self.pos < len(self.source) and (self.peek().isalnum() or self.peek() == "_"):
            value += self.advance()

        kind = KEYWORDS.get(value, TokenKind.IDENT)
        return Token(kind, value, Span(start_line, start_col, len(value)))

    def read_token(self) -> Token:
        self.skip_whitespace()

        while self.pos < len(self.source):
            if not self.skip_comment():
                break
            self.skip_whitespace()

        if self.pos >= len(self.source):
            return Token(TokenKind.EOF, "", Span(self.line, self.col))

        start_line, start_col = self.line, self.col
        ch = self.peek()

        if ch == '"':
            return self.read_string()
        if ch == "'":
            return self.read_char()
        if ch.isdigit():
            return self.read_number()
        if ch.isalpha() or ch == "_":
            return self.read_ident()

        self.advance()

        single = {
            "(": TokenKind.LPAREN,
            ")": TokenKind.RPAREN,
            "{": TokenKind.LBRACE,
            "}": TokenKind.RBRACE,
            "[": TokenKind.LBRACKET,
            "]": TokenKind.RBRACKET,
            ";": TokenKind.SEMICOLON,
            ",": TokenKind.COMMA,
            ".": TokenKind.DOT,
            "|": TokenKind.PIPE,
            "_": TokenKind.UNDERSCORE,
        }

        if ch in single:
            return Token(single[ch], ch, Span(start_line, start_col))

        if ch == ":":
            if self.peek(0) == ":":
                self.advance()
                return Token(TokenKind.DOUBLE_COLON, "::", Span(start_line, start_col, 2))
            return Token(TokenKind.COLON, ":", Span(start_line, start_col))

        if ch == "=":
            if self.peek(0) == ">":
                self.advance()
                return Token(TokenKind.FAT_ARROW, "=>", Span(start_line, start_col, 2))
            if self.peek(0) == "=":
                self.advance()
                return Token(TokenKind.EQ, "==", Span(start_line, start_col, 2))
            return Token(TokenKind.ASSIGN, "=", Span(start_line, start_col))

        if ch == "!":
            if self.peek(0) == "=":
                self.advance()
                return Token(TokenKind.NEQ, "!=", Span(start_line, start_col, 2))
            return Token(TokenKind.BANG, "!", Span(start_line, start_col))

        if ch == "<":
            if self.peek(0) == "=":
                self.advance()
                return Token(TokenKind.LTE, "<=", Span(start_line, start_col, 2))
            return Token(TokenKind.LT, "<", Span(start_line, start_col))

        if ch == ">":
            if self.peek(0) == "=":
                self.advance()
                return Token(TokenKind.GTE, ">=", Span(start_line, start_col, 2))
            return Token(TokenKind.GT, ">", Span(start_line, start_col))

        if ch == "+":
            if self.peek(0) == "=":
                self.advance()
                return Token(TokenKind.PLUS_ASSIGN, "+=", Span(start_line, start_col, 2))
            return Token(TokenKind.PLUS, "+", Span(start_line, start_col))

        if ch == "-":
            if self.peek(0) == "=":
                self.advance()
                return Token(TokenKind.MINUS_ASSIGN, "-=", Span(start_line, start_col, 2))
            if self.peek(0) == ">":
                self.advance()
                return Token(TokenKind.ARROW, "->", Span(start_line, start_col, 2))
            return Token(TokenKind.MINUS, "-", Span(start_line, start_col))

        if ch == "*":
            if self.peek(0) == "=":
                self.advance()
                return Token(TokenKind.STAR_ASSIGN, "*=", Span(start_line, start_col, 2))
            return Token(TokenKind.STAR, "*", Span(start_line, start_col))

        if ch == "/":
            if self.peek(0) == "=":
                self.advance()
                return Token(TokenKind.SLASH_ASSIGN, "/=", Span(start_line, start_col, 2))
            return Token(TokenKind.SLASH, "/", Span(start_line, start_col))

        if ch == "%":
            if self.peek(0) == "=":
                self.advance()
                return Token(TokenKind.PERCENT_ASSIGN, "%=", Span(start_line, start_col, 2))
            return Token(TokenKind.PERCENT, "%", Span(start_line, start_col))

        if ch == "&":
            if self.peek(0) == "&":
                self.advance()
                return Token(TokenKind.AND, "&&", Span(start_line, start_col, 2))
            return Token(TokenKind.AMP, "&", Span(start_line, start_col))

        if ch == "?":
            return Token(TokenKind.QUESTION, "?", Span(start_line, start_col))

        raise LexError(f"unexpected character: {ch!r}", Span(start_line, start_col), self.filename)

    def tokenize(self) -> List[Token]:
        while True:
            tok = self.read_token()
            self.tokens.append(tok)
            if tok.kind == TokenKind.EOF:
                break
        return self.tokens

File: compiler/src/lexer.rs
use crate::tokens::{Token, TokenKind};

pub struct Lexer {
    source: Vec<char>,
    filename: String,
    pos: usize,
    line: usize,
    column: usize,
}

impl Lexer {
    pub fn new(source: &str, filename: &str) -> Self {
        Lexer {
            source: source.chars().collect(),
            filename: filename.to_string(),
            pos: 0,
            line: 1,
            column: 1,
        }
    }

    pub fn tokenize(&mut self) -> Vec<Token> {
        let mut tokens = Vec::new();
        
        loop {
            let token = self.next_token();
            let is_eof = token.kind == TokenKind::Eof;
            tokens.push(token);
            if is_eof {
                break;
            }
        }
        
        tokens
    }

    fn next_token(&mut self) -> Token {
        self.skip_whitespace_and_comments();
        
        if self.pos >= self.source.len() {
            return Token::new(TokenKind::Eof, self.line, self.column);
        }

        let ch = self.source[self.pos];
        let start_line = self.line;
        let start_col = self.column;

        // String literals
        if ch == '"' {
            return self.read_string(start_line, start_col);
        }

        // Numbers
        if ch.is_ascii_digit() {
            return self.read_number(start_line, start_col);
        }

        // Identifiers and keywords
        if ch.is_ascii_alphabetic() || ch == '_' {
            return self.read_identifier(start_line, start_col);
        }

        // Two-character tokens
        if self.pos + 1 < self.source.len() {
            let next = self.source[self.pos + 1];
            let two_char = format!("{}{}", ch, next);
            
            match two_char.as_str() {
                "==" => { self.advance(2); return Token::new(TokenKind::Eq, start_line, start_col); }
                "!=" => { self.advance(2); return Token::new(TokenKind::Neq, start_line, start_col); }
                "<=" => { self.advance(2); return Token::new(TokenKind::Lte, start_line, start_col); }
                ">=" => { self.advance(2); return Token::new(TokenKind::Gte, start_line, start_col); }
                "&&" => { self.advance(2); return Token::new(TokenKind::And, start_line, start_col); }
                "||" => { self.advance(2); return Token::new(TokenKind::Or, start_line, start_col); }
                "->" => { self.advance(2); return Token::new(TokenKind::Arrow, start_line, start_col); }
                "::" => { self.advance(2); return Token::new(TokenKind::DoubleColon, start_line, start_col); }
                _ => {}
            }
        }

        // Single-character tokens
        self.advance(1);
        match ch {
            '+' => Token::new(TokenKind::Plus, start_line, start_col),
            '-' => Token::new(TokenKind::Minus, start_line, start_col),
            '*' => Token::new(TokenKind::Star, start_line, start_col),
            '/' => Token::new(TokenKind::Slash, start_line, start_col),
            '%' => Token::new(TokenKind::Percent, start_line, start_col),
            '<' => Token::new(TokenKind::Lt, start_line, start_col),
            '>' => Token::new(TokenKind::Gt, start_line, start_col),
            '!' => Token::new(TokenKind::Not, start_line, start_col),
            '=' => Token::new(TokenKind::Assign, start_line, start_col),
            '(' => Token::new(TokenKind::LParen, start_line, start_col),
            ')' => Token::new(TokenKind::RParen, start_line, start_col),
            '{' => Token::new(TokenKind::LBrace, start_line, start_col),
            '}' => Token::new(TokenKind::RBrace, start_line, start_col),
            '[' => Token::new(TokenKind::LBracket, start_line, start_col),
            ']' => Token::new(TokenKind::RBracket, start_line, start_col),
            ',' => Token::new(TokenKind::Comma, start_line, start_col),
            ':' => Token::new(TokenKind::Colon, start_line, start_col),
            ';' => Token::new(TokenKind::Semicolon, start_line, start_col),
            '.' => Token::new(TokenKind::Dot, start_line, start_col),
            '?' => Token::new(TokenKind::Question, start_line, start_col),
            _ => panic!("Unexpected character '{}' at {}:{}", ch, self.filename, start_line),
        }
    }

    fn skip_whitespace_and_comments(&mut self) {
        while self.pos < self.source.len() {
            let ch = self.source[self.pos];
            
            if ch == '\n' {
                self.line += 1;
                self.column = 1;
                self.pos += 1;
            } else if ch == '\r' {
                self.pos += 1;
                if self.pos < self.source.len() && self.source[self.pos] == '\n' {
                    self.pos += 1;
                }
                self.line += 1;
                self.column = 1;
            } else if ch == '\t' || ch == ' ' {
                self.column += 1;
                self.pos += 1;
            } else if ch == '/' && self.pos + 1 < self.source.len() && self.source[self.pos + 1] == '/' {
                // Line comment
                while self.pos < self.source.len() && self.source[self.pos] != '\n' {
                    self.pos += 1;
                }
            } else if ch == '/' && self.pos + 1 < self.source.len() && self.source[self.pos + 1] == '*' {
                // Block comment
                self.pos += 2;
                self.column += 2;
                while self.pos + 1 < self.source.len() {
                    if self.source[self.pos] == '*' && self.source[self.pos + 1] == '/' {
                        self.pos += 2;
                        self.column += 2;
                        break;
                    }
                    if self.source[self.pos] == '\n' {
                        self.line += 1;
                        self.column = 1;
                    } else {
                        self.column += 1;
                    }
                    self.pos += 1;
                }
            } else {
                break;
            }
        }
    }

    fn read_string(&mut self, start_line: usize, start_col: usize) -> Token {
        self.advance(1); // Skip opening quote
        let mut value = String::new();
        
        while self.pos < self.source.len() && self.source[self.pos] != '"' {
            if self.source[self.pos] == '\\' {
                self.advance(1);
                if self.pos < self.source.len() {
                    match self.source[self.pos] {
                        'n' => value.push('\n'),
                        't' => value.push('\t'),
                        '\\' => value.push('\\'),
                        '"' => value.push('"'),
                        _ => value.push(self.source[self.pos]),
                    }
                    self.advance(1);
                }
            } else {
                value.push(self.source[self.pos]);
                self.advance(1);
            }
        }
        
        if self.pos >= self.source.len() {
            panic!("Unterminated string at {}:{}", self.filename, start_line);
        }
        
        self.advance(1); // Skip closing quote
        Token::new(TokenKind::String(value), start_line, start_col)
    }

    fn read_number(&mut self, start_line: usize, start_col: usize) -> Token {
        let mut value = String::new();
        let mut is_float = false;
        
        while self.pos < self.source.len() && (self.source[self.pos].is_ascii_digit() || self.source[self.pos] == '.') {
            if self.source[self.pos] == '.' {
                if is_float {
                    break;
                }
                is_float = true;
            }
            value.push(self.source[self.pos]);
            self.advance(1);
        }
        
        if is_float {
            Token::new(TokenKind::Float(value.parse().unwrap()), start_line, start_col)
        } else {
            Token::new(TokenKind::Integer(value.parse().unwrap()), start_line, start_col)
        }
    }

    fn read_identifier(&mut self, start_line: usize, start_col: usize) -> Token {
        let mut value = String::new();
        
        while self.pos < self.source.len() && (self.source[self.pos].is_ascii_alphanumeric() || self.source[self.pos] == '_') {
            value.push(self.source[self.pos]);
            self.advance(1);
        }
        
        let kind = match value.as_str() {
            "fn" => TokenKind::Fn,
            "let" => TokenKind::Let,
            "mut" => TokenKind::Mut,
            "if" => TokenKind::If,
            "else" => TokenKind::Else,
            "while" => TokenKind::While,
            "return" => TokenKind::Return,
            "print" => TokenKind::Print,
            "true" => TokenKind::Bool(true),
            "false" => TokenKind::Bool(false),
            "struct" => TokenKind::Struct,
            "enum" => TokenKind::Enum,
            "domain" => TokenKind::Domain,
            "spawn" => TokenKind::Spawn,
            "await" => TokenKind::Await,
            "chan" => TokenKind::Chan,
            "new" => TokenKind::New,
            "Some" => TokenKind::Some,
            "None" => TokenKind::None,
            "maybe" => TokenKind::Maybe,
            "type" => TokenKind::Type,
            "as" => TokenKind::As,
            "in" => TokenKind::In,
            "loop" => TokenKind::Loop,
            "break" => TokenKind::Break,
            "continue" => TokenKind::Continue,
            _ => TokenKind::Identifier(value),
        };
        
        Token::new(kind, start_line, start_col)
    }

    fn advance(&mut self, count: usize) {
        for _ in 0..count {
            if self.pos < self.source.len() {
                if self.source[self.pos] == '\n' {
                    self.line += 1;
                    self.column = 1;
                } else {
                    self.column += 1;
                }
                self.pos += 1;
            }
        }
    }
}
